Geometry/CommonTopologies/src/GEMStripTopology.cc

```diff
@@ -0,0 +1,125 @@
+
+      T(angularWidth() * (centreToIntersection() + yAxisOrientation() * mp.y() * detHeight()) /
+        c1),                // tangential measurement unit (local pitch)
+      R(detHeight() / c1),  // radial measurement unit (strip length)
+      tt(me.uu() * T * T),  // tangential sigma^2
+      rr(me.vv() * R * R),  // radial sigma^2
+      tr(me.uv() * T * R),
+
+      xx(c2 * tt + 2 * cs * tr + s2 * rr), yy(s2 * tt - 2 * cs * tr + c2 * rr), xy(cs * (rr - tt) + tr * (c2 - s2));
+
+  return LocalError(xx, xy, yy);
+}
+
+float GEMStripTopology::strip(const LocalPoint& lp) const {
+  const float  // phi is measured from y axis --> sign of angle is sign of x * yAxisOrientation --> use atan2(x,y), not atan2(y,x)
+      phi(std::atan2(lp.x(), yDistanceToIntersection(lp.y()))),
+      aStrip((phi - yAxisOrientation() * phiOfOneEdge()) / angularWidth());
+  return std::max(float(0), std::min((float)nstrips(), aStrip));
+}
+
+int GEMStripTopology::nearestStrip(const LocalPoint& lp) const {
+  return std::min(nstrips(), static_cast<int>(std::max(float(0), strip(lp))) + 1);
+}
+
+MeasurementPoint GEMStripTopology::measurementPosition(const LocalPoint& lp) const {
+  const float  // phi is [pi/2 - conventional local phi], use atan2(x,y) rather than atan2(y,x)
+      phi(yAxisOrientation() * std::atan2(lp.x(), yDistanceToIntersection(lp.y())));
+  return MeasurementPoint(yAxisOrientation() * (phi - phiOfOneEdge()) / angularWidth(),
+                          (lp.y() - yCentreOfStripPlane()) / detHeight());
+}
+
+MeasurementError GEMStripTopology::measurementError(const LocalPoint& p, const LocalError& e) const {
+  const double yHitToInter(yDistanceToIntersection(p.y())),
+      t(yAxisOrientation() * p.x() / yHitToInter),  // tan(strip angle)
+      cs(t / (1 + t * t)), s2(t * cs), c2(1 - s2),  // rotation matrix
+
+      T2(1. / (std::pow(angularWidth(), 2) *
+               (std::pow(p.x(), 2) + std::pow(yHitToInter, 2)))),  // 1./tangential measurement unit (local pitch) ^2
+      R2(c2 / std::pow(detHeight(), 2)),                           // 1./ radial measurement unit (strip length) ^2
+
+      uu((c2 * e.xx() - 2 * cs * e.xy() + s2 * e.yy()) * T2), vv((s2 * e.xx() + 2 * cs * e.xy() + c2 * e.yy()) * R2),
+      uv((cs * (e.xx() - e.yy()) + e.xy() * (c2 - s2)) * std::sqrt(T2 * R2));
+
+  return MeasurementError(uu, uv, vv);
+}
+
+int GEMStripTopology::channel(const LocalPoint& lp) const { return std::min(int(strip(lp)), numberOfStrips_ - 1); }
+
+float GEMStripTopology::pitch() const { return localPitch(LocalPoint(0, 0)); }
+
+float GEMStripTopology::localPitch(const LocalPoint& lp) const {
+  const int istrip = std::min(nstrips(), static_cast<int>(strip(lp)) + 1);  // which strip number
+  const float fangle = stripAngle(static_cast<float>(istrip) - 0.5);        // angle of strip centre
+  return yDistanceToIntersection(lp.y()) * std::sin(angularWidth()) /
+         std::pow(std::cos(fangle - 0.5f * angularWidth()), 2);
```

Comment:
I think there is a possibility of division by 0 here also, which should be prevented.

Reply:
yes, I added the assert function.